flamedisx/lxe_sources.py

```diff
@@ -0,0 +1,294 @@
+        del wimp_kwargs['energy_edges']
+        spectra = np.array([wr.rate_wimp_std(t=t,
+                                             es=e_centers,
+                                             **wimp_kwargs)
+                            * np.diff(self.energy_edges)
+                            for t in time_centers])
+        assert spectra.shape == (len(time_centers), len(e_centers))
+
+        self.energy_hist = Histdd.from_histogram(
+            spectra,
+            bin_edges=(times, self.energy_edges))
+
+        if self.pretend_wimps_dont_modulate:
+            self.energy_hist.histogram = (
+                np.ones_like(self.energy_hist.histogram)
+                * self.energy_hist.sum(axis=0).histogram.reshape(1, -1)
+                / self.n_time_bins)
+
+        super().__init__(*args, **kwargs)
+
+    def energy_spectrum(self, event_time):
+        t_j2000 = wr.j2000(fd.tf_to_np(event_time))
+        result = np.stack([self.energy_hist.slicesum(t).histogram
+                           for t in t_j2000])
+        return fd.np_to_tf(result)
+
+    def mu_before_efficiencies(self, **params):
+        return self.energy_hist.n / self.n_time_bins
+
+    def random_truth(self, n_events, fix_truth=None, **params):
+        """Draw n_events random energies and times from the energy/
+        time spectrum and add them to the data dict.
+        """
+        data = self.draw_positions(n_events)
+
+        if 'event_time' in fix_truth:
+            # Time is fixed, so the energy spectrum differs.
+            # (if energy is also fixed, it will just be overridden later
+            #  and we're doing a bit of unnecessary work here)
+            t = wr.j2000(fix_truth['event_time'])
+            t_edges = self.energy_hist.bin_edges[0]
+            assert t_edges[0] <= t < t_edges[-1], "fix_truth time out of bounds"
+            data['energy'] = \
+                self.energy_hist \
+                    .slicesum(t, axis=0) \
+                    .get_random(n_events)
+
+        elif 'energy' in fix_truth:
+            # Energy is fixed, so the time distribution differs.
+            e_edges = self.energy_hist.bin_edges[1]
+            assert e_edges[0] <= fix_truth['energy'] < e_edges[-1], \
+                "fix_truth energy out of bounds"
+            times = \
+                self.energy_hist \
+                    .slicesum(fix_truth['energy'], axis=1) \
+                    .get_random(n_events)
+            data['event_time'] = fd.j2000_to_event_time(times)
+
+        else:
+            times, data['energy'] = self.energy_hist.get_random(n_events).T
+            data['event_time'] = fd.j2000_to_event_time(times)
+
+        # Overwrite all fixed truth keys
+        self._overwrite_fixed_truths(data, fix_truth, n_events)
+
+        return pd.DataFrame(data)
+
+    @staticmethod
+    def bin_centers(x):
+        return 0.5 * (x[1:] + x[:-1])
+
+
+@export
+class SpatialRateHistogramSource(fd.BlockModelSource):
```

Comment:
Good point, this would not work right; in the tests we copy over the blocks, but that ignores any model functions... I'll see if I can make SpatialRateHistogram an energy spectrum block.

Reply:
Made a SpatialRateEnergySpectrum and WIMPEnergySpectrum block now